[#2118] Implement keyword snippets for control-flow keywords (if/for/while/etc.)

DPrakashh opened the pull request:
This PR implements snippet completions for control-flow keywords (if, for, while, guard, switch, repeat) when they are used in code completion.

The solution involved updating "CodeCompletionSession.swift" to identify keyword completions (.keyword) and overwrite their "insertText" and "textEdit" with the correct LSP snippet strings, setting the insertTextFormat to .snippet.

This ensures that keywords now contain placeholders (e.g., `${1:condition}` and `$0`) required by LSP clients.

Verification
New unit tests were added to "SwiftCompletionSnippetTests.swift" to explicitly verify:
1. The "insertTextFormat" is correctly set to ".snippet".
2. The presence of the required snippet placeholders (e.g., `$1`, `$0`).

All local tests, including the focused test (`swift test --filter SwiftCompletionSnippetTests/testKeywordIfProvidesSnippet`), passed successfully.

Checklist (Based on Contributing Guide)
1-Code has been formatted using `swift format -ipr 
2-New tests were added to cover the feature.

Fixes #2118

DPrakashh commented:
Hi @ahoppen , @bnbarham , @hamishknight , @rintaro 

Hope you're having a good week.

I wanted to check on the status of my PR from four days ago. I know these core completion changes take time to review carefully.
Just confirming that all local tests passed on my end. Would it be possible to get the CI tests triggered when a maintainer is able to look at it?

Thanks for your time!

ahoppen commented on Sources/SwiftLanguageService/CodeCompletionSession.swift:
```diff
@@ -704,6 +719,29 @@ class CodeCompletionSession {
+
+    switch keyword {
+    case "if":
+      return "if ${1:condition} {\n\t${0:}\n}"
```
We should not assume that the source file is indented with tabs. We have `BasicFormat.inferIndenatation` to infer a source files’s indentation and should rely and that instead.

ahoppen commented on Tests/SourceKitLSPTests/SwiftCompletetionSnippetTests.swift:
```diff
@@ -0,0 +1,247 @@
+@_spi(SourceKitLSP) import LanguageServerProtocol
```
Could you add the copyright header to this file?

ahoppen reviewed (commented):
Thank you. The overall structure looks good, I have left a few comments inline. 

Also, do you have a screen recording or something like that how this looks like in VS Code? Ie. point VS Code to a locally built version of SourceKit-LSP as described in https://github.com/swiftlang/sourcekit-lsp/blob/main/CONTRIBUTING.md#using-a-locally-built-sourcekit-lsp-in-an-editor.

ahoppen commented on Sources/SwiftLanguageService/CodeCompletionSession.swift:
```diff
@@ -704,6 +714,50 @@ class CodeCompletionSession {
+
+    func indentationUnitString() -> String {
+      // Convert the inferred Trivia (spaces/tabs) to a string we can embed into snippets.
+      if let trivia = indentationWidth {
```
Nice, I forgot that we already have the inferred indentation in `CodeCompletionSession` 🤩

ahoppen commented on Sources/SwiftLanguageService/CodeCompletionSession.swift:
```diff
@@ -704,6 +714,50 @@ class CodeCompletionSession {
+    case "guard":
+      return "guard ${1:condition} else {\n\(indent)${0:}\n}"
+    case "switch":
+      return "switch ${1:value} {\n\(indent)case ${2:pattern}:\n\(doubleIndent)${0:}\n}"
```
Let’s not indent the `case` label to match the default behavior of [swift-format](https://github.com/swiftlang/swift-format/blob/main/Documentation/Configuration.md#indentswitchcaselabels)

Ie. make this

```suggestion
      return """
        switch ${1:value} {
        case ${2:pattern}:
        \(indent)${0:}
        }
        """
```

ahoppen commented on Sources/SwiftLanguageService/CodeCompletionSession.swift:
```diff
@@ -704,6 +714,50 @@ class CodeCompletionSession {
+
+    switch keyword {
+    case "if":
+      return "if ${1:condition} {\n\(indent)${0:}\n}"
```
Let’s call the 0 placeholder `body`, ie. `${0:body}`.

DPrakashh commented:
@ahoppen 

I have finalised the refinements based on your feedback. I am attaching a video demo (recorded with Screencast Mode) which demonstrates the comparison between the default VS Code snippets and the updated SourceKit-LSP snippets.

**Changes implemented based on your suggestions:**
1-Logical Tab Order: Reordered the tab stops so the cursor lands on the condition first ($1) and moves to the body last ($0). This ensures the user fills out the logic before jumping into the braces.
2-Switch Statement Alignment: Corrected the switch snippet so that case labels align vertically with the switch keyword, fixing the previous double-indentation issue.
3-Code Cleanup: Removed the unused doubleIndent variable and the legacy indentation helper functions, opting for a cleaner implementation using indentationWidth.
4-Refined Test Suite: Updated CodeCompletionSnippetTests to be more resilient, verifying the presence of specific keywords and placeholders rather than exact indentation strings.

**Usability Improvement:**
1-**Descriptive "Body" Placeholder:** I added a small improvement for the user by changing the final stop from an empty ${0:} to a descriptive ${0:body}. This provides a clear visual hint of where the code should go while remaining the final exit point for the snippet.

I have verified these changes using a locally built SourceKit-LSP server in VS Code as shown in the video. Ready for your final review!

**Note**: I’ve done my best to address all the points from the review. If there are any further refinements or "nits" you'd like me to fix to get this fully aligned with the project's standards, please let me know—I’m happy to make additional changes!
https://github.com/user-attachments/assets/f14026c9-c66b-4c39-b836-8aa5dcbdd270

ahoppen reviewed (approved):
Looks great to me. Thank you @DPrakashh ❤️

ahoppen commented:
@swift-ci Please test Windows
